orttraining/orttraining/python/training/ortmodule.py

```diff
@@ -76,6 +83,62 @@ def _onnx_value_info_to_buffer_tensor(value_info, device):
     dtype = _utils.dtype_onnx_to_torch(value_info.type.tensor_type.elem_type)
     return torch.zeros(shape, device=device, dtype=dtype)
 
+def _parse_inputs_for_onnx_export(module, inputs):
+    # Ignore optional *inputs explicitly specified as None
+    sig = signature(module.forward)
+    all_input_names = sig.parameters.keys()
+    input_names = []
+    dynamic_axes = {}
+    for input_idx, name in enumerate(all_input_names):
+        if input_idx < len(inputs) and inputs[input_idx] is not None:
+            input_names.append(name)
+            dynamic_axes[name] = {}
+            for dim_idx in range(len(inputs[input_idx].shape)):
+                dynamic_axes[name].update({dim_idx : 'input{}_dim{}'.format(input_idx, dim_idx)})
+    return input_names, dynamic_axes
+
+def _parse_outputs_for_onnx_export(module, inputs):
+    #   Do an inference to grab outputs
+    is_train_mode = module.training
+    module.eval()
+    with torch.no_grad():
+        # Deepcopy inputs, since input values may change after model run.
+        sample_inputs_copy = _deepcopy_model_input(*inputs)
+        try:
+            # Deepcopy model, in case model is stateful and changes after model run.
+            model_copy = copy.deepcopy(module)
+        except Exception:
+            model_copy = module
+            warnings.warn("This model cannot be deep copied (or pickled), which is a required step for stateful models to be properly exported to ONNX."
+                            " Compute will continue, but unexpected results may occur!")
+
+        sample_outputs = model_copy(*sample_inputs_copy)
+        def _create_output_dim_names(output, output_idx):
```

Comment:
Any reason this function is defined inside the ```torch.no_grad()``` context? May we move it outside or at the beginning of this function?

Reply:
Reasoning was defining it in the most specific context, close to the rest of that block